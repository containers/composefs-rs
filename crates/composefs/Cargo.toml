[package]
name = "composefs"
description = "Rust library for the composefs filesystem"
keywords = ["composefs"]

edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[features]
'pre-6.15' = ['tempfile']
rhel9 = ['pre-6.15']
test = []

[dependencies]
anyhow = { version = "1.0.87", default-features = false }
hex = { version = "0.4.0", default-features = false, features = ["std"] }
log = { version = "0.4.8", default-features = false }
once_cell = { version = "1.21.3", default-features = false, features = ["std"] }
rustix = { version = "1.0.0", default-features = false, features = ["fs", "mount", "process", "std"] }
sha2 = { version = "0.10.1", default-features = false, features = ["std"] }
thiserror = { version = "2.0.0", default-features = false }
tokio = { version = "1.24.2", default-features = false, features = ["io-util", "rt-multi-thread"] }
tempfile = { version = "3.8.0", optional = true, default-features = false }
xxhash-rust = { version = "0.8.2", default-features = false, features = ["xxh32"] }
zerocopy = { version = "0.8.0", default-features = false, features = ["derive", "std"] }
zstd = { version = "0.13.0", default-features = false }
rand = { version = "0.9.1", default-features = true }

[dev-dependencies]
insta = "1.42.2"
similar-asserts = "1.7.0"
tempfile = { version = "3.8.0", default-features = false }
test-with = { version = "0.14", default-features = false, features = ["executable", "runtime"] }
tokio-test = "0.4.4"

[lints]
workspace = true
