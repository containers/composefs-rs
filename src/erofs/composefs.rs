use zerocopy::{FromBytes, Immutable, IntoBytes, KnownLayout};

use crate::fsverity::FsVerityHashValue;

/* From linux/fs/overlayfs/overlayfs.h struct ovl_metacopy */
#[derive(Debug, FromBytes, Immutable, KnownLayout, IntoBytes)]
#[repr(C)]
pub(super) struct OverlayMetacopy<H: FsVerityHashValue> {
    version: u8,
    len: u8,
    flags: u8,
    digest_algo: u8,
    pub(super) digest: H,
}

impl<H: FsVerityHashValue> OverlayMetacopy<H> {
    pub(super) fn new(digest: &H) -> Self {
        Self {
            version: 0,
            len: size_of::<Self>() as u8,
            flags: 0,
            digest_algo: H::ALGORITHM,
            digest: digest.clone(),
        }
    }

    pub(super) fn valid(&self) -> bool {
        self.version == 0
            && self.len == size_of::<Self>() as u8
            && self.flags == 0
            && self.digest_algo == H::ALGORITHM
    }
}
