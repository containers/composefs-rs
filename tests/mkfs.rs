use std::{
    cell::RefCell,
    collections::BTreeMap,
    ffi::OsStr,
    io::Write,
    process::{Command, Stdio},
    rc::Rc,
};

use similar_asserts::assert_eq;
use tempfile::NamedTempFile;

use composefs::{
    dumpfile::write_dumpfile,
    erofs::{debug::debug_img, writer::mkfs_erofs},
    fsverity::{FsVerityHashValue, Sha256HashValue},
    image::{Directory, FileSystem, Inode, Leaf, LeafContent, RegularFile, Stat},
};

fn debug_fs(mut fs: FileSystem) -> String {
    fs.done();
    let image = mkfs_erofs(&fs);
    let mut output = vec![];
    debug_img(&mut output, &image).unwrap();
    String::from_utf8(output).unwrap()
}

fn empty(_fs: &mut FileSystem) {}

#[test]
fn test_empty() {
    let mut fs = FileSystem::new();
    empty(&mut fs);
    insta::assert_snapshot!(debug_fs(fs));
}

fn add_leaf(dir: &mut Directory, name: impl AsRef<OsStr>, content: LeafContent) {
    dir.insert(
        name.as_ref(),
        Inode::Leaf(Rc::new(Leaf {
            content,
            stat: Stat {
                st_gid: 0,
                st_uid: 0,
                st_mode: 0,
                st_mtim_sec: 0,
                xattrs: RefCell::new(BTreeMap::new()),
            },
        })),
    );
}

fn simple(fs: &mut FileSystem) {
    let ext_id = Sha256HashValue::from_hex(
        "5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a",
    )
    .unwrap();
    add_leaf(&mut fs.root, "fifo", LeafContent::Fifo);
    add_leaf(
        &mut fs.root,
        "regular-inline",
        LeafContent::Regular(RegularFile::Inline((*b"hihi").into())),
    );
    add_leaf(
        &mut fs.root,
        "regular-external",
        LeafContent::Regular(RegularFile::External(ext_id, 1234)),
    );
    add_leaf(&mut fs.root, "chrdev", LeafContent::CharacterDevice(123));
    add_leaf(&mut fs.root, "blkdev", LeafContent::BlockDevice(123));
    add_leaf(&mut fs.root, "socket", LeafContent::Socket);
    add_leaf(
        &mut fs.root,
        "symlink",
        LeafContent::Symlink(OsStr::new("/target").into()),
    );
}

#[test]
fn test_simple() {
    let mut fs = FileSystem::new();
    simple(&mut fs);
    insta::assert_snapshot!(debug_fs(fs));
}

fn foreach_case(f: fn(&FileSystem)) {
    for case in [empty, simple] {
        let mut fs = FileSystem::new();
        case(&mut fs);
        fs.done();
        f(&fs);
    }
}

#[test_with::executable(fsck.erofs)]
fn test_fsck() {
    foreach_case(|fs| {
        let mut tmp = NamedTempFile::new().unwrap();
        tmp.write_all(&mkfs_erofs(fs)).unwrap();
        let mut fsck = Command::new("fsck.erofs").arg(tmp.path()).spawn().unwrap();
        assert!(fsck.wait().unwrap().success());
    });
}

fn dump_image(img: &[u8]) -> String {
    let mut dump = vec![];
    debug_img(&mut dump, img).unwrap();
    String::from_utf8(dump).unwrap()
}

#[should_panic]
#[test_with::executable(mkcomposefs)]
fn test_vs_mkcomposefs() {
    foreach_case(|fs| {
        let image = mkfs_erofs(fs);

        let mut mkcomposefs = Command::new("mkcomposefs")
            .args(["--min-version=3", "--from-file", "-", "-"])
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()
            .unwrap();

        let mut stdin = mkcomposefs.stdin.take().unwrap();
        write_dumpfile(&mut stdin, fs).unwrap();
        drop(stdin);

        let output = mkcomposefs.wait_with_output().unwrap();
        assert!(output.status.success());
        let mkcomposefs_image = output.stdout.into_boxed_slice();

        if image != mkcomposefs_image {
            let dump = dump_image(&image);
            let mkcomposefs_dump = dump_image(&mkcomposefs_image);
            assert_eq!(mkcomposefs_dump, dump);
        }
        assert_eq!(image, mkcomposefs_image); // fallback if the dump is somehow the same
    });
}
