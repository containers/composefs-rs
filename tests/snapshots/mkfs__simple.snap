---
source: tests/mkfs.rs
expression: debug_fs(fs)
snapshot_kind: text
---
00000000 ComposefsHeader
      +0     magic: U32(3497550490)
      +4     version: U32(1)
      +c     composefs_version: U32(2)

00000020 Padding
    +3e0     # 992 nul bytes

00000400 Superblock
      +0     magic: U32(3774210530)
      +8     feature_compat: U32(6)
      +c     blkszbits: 12
      +e     root_nid: U16(36)
     +10     inos: U64(8)
     +24     blocks: U32(1)

# Filename "/"
# nid #36
00000480 ExtendedInodeHeader
      +0     format: 5 = Extended | Ok(FlatInline)
      +4     mode: 0040555 (directory)
      +8     size: U64(170)
     +14     ino: U32(36)
     +2c     nlink: U32(2)
     +40 --- inline directory entries ---
      +0     inode_offset: U64(36)
      +8     name_offset: U16(108)
      +a     file_type: Directory
     +6c     # name: "."

      +c     inode_offset: U64(36)
     +14     name_offset: U16(109)
     +16     file_type: Directory
     +6d     # name: ".."

     +18     inode_offset: U64(44)
     +20     name_offset: U16(111)
     +22     file_type: BlockDevice
     +6f     # name: "blkdev"

     +24     inode_offset: U64(46)
     +2c     name_offset: U16(117)
     +2e     file_type: CharacterDevice
     +75     # name: "chrdev"

     +30     inode_offset: U64(48)
     +38     name_offset: U16(123)
     +3a     file_type: Fifo
     +7b     # name: "fifo"

     +3c     inode_offset: U64(50)
     +44     name_offset: U16(127)
     +46     file_type: RegularFile
     +7f     # name: "regular-external"

     +48     inode_offset: U64(57)
     +50     name_offset: U16(143)
     +52     file_type: RegularFile
     +8f     # name: "regular-inline"

     +54     inode_offset: U64(60)
     +5c     name_offset: U16(157)
     +5e     file_type: Socket
     +9d     # name: "socket"

     +60     inode_offset: U64(62)
     +68     name_offset: U16(163)
     +6a     file_type: Symlink
     +a3     # name: "symlink"

0000056a Padding
     +16     # 22 nul bytes

# Filename "/blkdev"
# nid #44
00000580 ExtendedInodeHeader
      +0     format: 1 = Extended | Ok(FlatPlain)
      +4     mode: 0060000 (blockdev)
     +10     u: U32(123)
     +14     ino: U32(44)
     +2c     nlink: U32(1)

# Filename "/chrdev"
# nid #46
000005c0 ExtendedInodeHeader
      +0     format: 1 = Extended | Ok(FlatPlain)
      +4     mode: 0020000 (chardev)
     +10     u: U32(123)
     +14     ino: U32(46)
     +2c     nlink: U32(1)

# Filename "/fifo"
# nid #48
00000600 ExtendedInodeHeader
      +0     format: 1 = Extended | Ok(FlatPlain)
      +4     mode: 0010000 (fifo)
     +14     ino: U32(48)
     +2c     nlink: U32(1)

# Filename "/regular-external"
# nid #50
00000640 ExtendedInodeHeader
      +0     format: 9 = Extended | Ok(ChunkBased)
      +2     xattr_icount: U16(37)
      +4     mode: 0100000 (regular file)
      +8     size: U64(1234)
     +10     u: U32(31)
     +14     ino: U32(50)
     +2c     nlink: U32(1)
     +40     name_filter: U32(2147352575)
     +4c     xattr: (4 16 36) trusted."overlay.metacopy" = "\0$\0\u{1}ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ"
     +84     xattr: (4 16 66) trusted."overlay.redirect" = "/5a/5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a"
     +d0                                       ff ff ff ff  |            ....|

# Filename "/regular-inline"
# nid #57
00000720 ExtendedInodeHeader
      +0     format: 5 = Extended | Ok(FlatInline)
      +4     mode: 0100000 (regular file)
      +8     size: U64(4)
     +14     ino: U32(57)
     +2c     nlink: U32(1)
     +40     inline: "hihi"

00000764 Padding
     +1c     # 28 nul bytes

# Filename "/socket"
# nid #60
00000780 ExtendedInodeHeader
      +0     format: 1 = Extended | Ok(FlatPlain)
      +4     mode: 0140000 (socket)
     +14     ino: U32(60)
     +2c     nlink: U32(1)

# Filename "/symlink"
# nid #62
000007c0 ExtendedInodeHeader
      +0     format: 5 = Extended | Ok(FlatInline)
      +4     mode: 0120000 (symlink)
      +8     size: U64(7)
     +14     ino: U32(62)
     +2c     nlink: U32(1)
     +40     inline: "/target"

00000807 Padding
    +7f9     # 2041 nul bytes


Space statistics (total size 4096B):
  extended inode = 853B, 20.83%
  header = 32B, 0.78%
  superblock = 128B, 3.12%
  padding extended inode -> eof = 2041B, 49.83%
  padding extended inode -> extended inode = 50B, 1.22%
  padding header -> superblock = 992B, 24.22%
